src/kakarot/stack.cairo

```diff
@@ -4,68 +4,82 @@
         alloc_locals;
-        let (elements: Uint256*) = alloc();
-        return new model.Stack(elements=elements, raw_len=0);
+        let (stack_word_dict_start: DictAccess*) = default_dict_new(0);
+        return new model.Stack(
+            stack_word_dict_start=stack_word_dict_start,
+            stack_word_dict=stack_word_dict_start,
+            stack_16bytes_len=0);
     }
 
-    // @notice Returns the length of the stack.
-    // @param self - The pointer to the stack.
-    // @return The length of the stack.
-    func len(self: model.Stack*) -> felt {
-        let actual_len = self.raw_len / element_size;
-        return actual_len;
+    // @notice Finalizes the stack.
+    // @return The pointer to the stack Summary.
+    func finalize{range_check_ptr}(self: model.Stack*) -> Summary* {
+        let (squashed_start, squashed_end) = default_dict_finalize(
+            self.stack_word_dict_start, self.stack_word_dict, 0
+        );
+        return new Summary(
+            stack_16bytes_len=self.stack_16bytes_len, stack_squashed_start=squashed_start, stack_squashed_end=squashed_end
+            );
     }
 
-    // @notice Push an element to the stack.
+    // @notice Store an element into the stack.
     // @param self - The pointer to the stack.
     // @param element - The element to push.
+    // @param offset - The offset to store the element at.
     // @return The new pointer to the stack.
     func push{range_check_ptr}(self: model.Stack*, element: Uint256) -> model.Stack* {
-        alloc_locals;
-        Stack.check_overflow(self=self);
-        assert [self.elements + self.raw_len] = element;
-        return new model.Stack(elements=self.elements, raw_len=self.raw_len + element_size);
-    }
+        let stack_word_dict = self.stack_word_dict;
+        let position_zero = self.stack_16bytes_len;
 
-    // @notice Pop an element from the stack.
-    // @param self - The pointer to the stack.
-    // @return The new pointer to the stack.
-    // @return The popped element.
-    func pop{range_check_ptr}(self: model.Stack*) -> (new_stack: model.Stack*, element: Uint256) {
-        alloc_locals;
-        Stack.check_underflow(self=self, stack_index=0);
-        // Get last element
-        let len = Stack.len(self=self);
-        let element = self.elements[len - 1];
-        // Get new segment for next stack copy
-        let (new_elements: Uint256*) = alloc();
-        // Get length of new stack copy
-        let new_len = self.raw_len - element_size;
-        // Copy stack without last element
-        memcpy(dst=new_elements, src=self.elements, len=new_len);
-        // Create new stack
-        local new_stack: model.Stack* = new model.Stack(elements=new_elements, raw_len=new_len);
-        return (new_stack=new_stack, element=element);
+        if (position_zero == Constants.STACK_MAX_DEPTH * 2 + 2) {
+            with_attr error_message("Kakarot: StackOverflow") {
+                assert 1 = 0;
```

Comment:
why not writing the actual check in the `with_attr`, smth like

```
with_attr error_message(...)
    assert_le(position_zero, Constants.STACK_MAX_DEPTH * 2)
```

Reply:
No need for assert_le, it's less efficient.
This check is efficient, see my answer below.